### PR TITLE
feat: automated purchase auctions

### LINKED ISSUE #11685: [Feature]: Automated Purchase Auctions
In order to allow the protocol to better manage collected fees a framework for regular, fully automated exchanges of one token for another on the protocol's spot markets should be implemented. In order to allow the protocol to receive a fair rate for the token exchange, and taking into account that not all spot markets will have deep liquidity, this should be handled through use of a scheduled auction.

This auction should have several defining features:

 - It should be triggerable on a defined schedule
 - An event should be emitted at a known time in advance of the auction announcing the amount to be exchanged
 - The price at which the protocol is willing to exchange should be configurable as some offset to a predefined set of oracle inputs

spec in https://github.com/vegaprotocol/specs/blob/805e231fb672314342a2e315a54a07de12bad60e/protocol/0093-PAPU-protocol_automated_purchases.md#protocol-automated-purchases

### PR DESCRIPTION
**Pull request by ze97286**
Closes #11685 
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1548
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1549
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1546
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1550
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1547
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1552
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1553
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1554
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1537
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1538
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1539
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1540
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1541
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1542
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1543
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1544
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1532
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1533
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1534
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1535
Closes https://github.com/vegaprotocol/core-test-coverage/issues/1536

### CODE REVIEW

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -7839,6 +7847,31 @@ enum AMMStatusReason {
...
+  from: ID!
+  "The account type for the network from which the tokens will be sourced"
+  fromAccountType: AccountType!
+  "The account type for the network to which the purchased tokens will be sent"
+  toAccountType: AccountType!
+  "The market which will be used to enact the purchase/sale"
+  marketId: ID
+  "The oracle which will define an approximate acceptable price for the transaction"
+  priceOracle: DataSourceDefinition!
+  "The final acceptable price should be the Price Oracle value weighted by this factor (e.g. 1.05 to allow for 5% slippage on the purcase)"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:7861]:
```suggestion
  "The final acceptable price is the price oracle value weighted by this factor (e.g. 1.05 to allow for 5% slippage on the purchase)"
```

File: protos/sources/data-node/api/v2/trading_data.proto [hunk 2]
```diff
@@ -3220,6 +3220,8 @@ message ListGovernanceDataRequest {
     TYPE_UPDATE_REFERRAL_PROGRAM = 13;
     // Proposal to update the volume discount program
     TYPE_UPDATE_VOLUME_DISCOUNT_PROGRAM = 14;
+    // Proposal foor new automated purchase
```
Comment (candida-d) [inline protos/sources/data-node/api/v2/trading_data.proto:3223]:
```suggestion
    // Proposal for new automated purchase auction
```

File: protos/sources/vega/events/v1/events.proto [hunk 3]
```diff
@@ -1546,6 +1546,9 @@ enum BusEventType {
   // Event indicating the updated statistics for the volume rebate.
   BUS_EVENT_TYPE_VOLUME_REBATE_STATS_UPDATED = 95;
 
+  // Event indicating the announcement of an automated purchase auction scheduled.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1549]:
```suggestion
  // Event indicating an automated purchase auction has been scheduled.
```

File: protos/sources/vega/governance.proto [hunk 4]
```diff
@@ -386,6 +386,8 @@ message ProposalTerms {
     UpdateVolumeDiscountProgram update_volume_discount_program = 113;
     // Proposal change for updating the volume rebate program.
     UpdateVolumeRebateProgram update_volume_rebate_program = 114;
+    // Proposal for a new buy back configuration
```
Comment (candida-d) [inline protos/sources/vega/governance.proto:389]:
```suggestion
    // Proposal for a new automated purchase auction
```

File: protos/sources/vega/governance.proto [hunk 5]
```diff
@@ -428,6 +430,8 @@ message BatchProposalTermsChange {
     NewAsset new_asset = 113;
     // Proposal change for updating the volume rebate program.
     UpdateVolumeRebateProgram update_volume_rebate_program = 114;
+    // Proposal for a new buy back configuration
```
Comment (candida-d) [inline protos/sources/vega/governance.proto:433]:
```suggestion
    // Proposal for a new automated purchase auction
```

File: protos/sources/vega/governance.proto [hunk 6]
```diff
@@ -882,3 +888,37 @@ message RecurringTransfer {
   // For example if the initial transfer amount is 1000 and the factor is 0.5, then the amounts transferred per epoch will be 1000, 500, 250, 125, etc.
   string factor = 4;
 }
+
+message NewAutomatedPurchase {
+  // Configuration for a change to create a new automated purchase
```
Comment (candida-d) [inline protos/sources/vega/governance.proto:893]:
```suggestion
  // Configuration for a change to create a new automated purchase auction
```

File: protos/sources/vega/governance.proto [hunk 7]
```diff
@@ -882,3 +888,37 @@ message RecurringTransfer {
...
+  vega.AccountType from_account_type = 2;
+  // The account type for the network to which the purchased tokens will be sent
+  vega.AccountType to_account_type = 3;
+  // The market which will be used to enact the purchase/sale
+  string market_id = 4;
+  // The oracle which will define an approximate acceptable price for the transaction
+  vega.DataSourceDefinition price_oracle = 5;
+  // The spec binding for the price oracle
+  vega.SpecBindingForCompositePrice price_oracle_spec_binding = 6;
+  // The final acceptable price should be the Price Oracle value weighted by this factor (e.g. 1.05 to allow for 5% slippage on the purcase)
```
Comment (candida-d) [inline protos/sources/vega/governance.proto:910]:
```suggestion
  // The final acceptable price is the price oracle value weighted by this factor (e.g. 1.05 to allow for 5% slippage on the purchase)
```

File: protos/sources/vega/markets.proto [hunk 8]
```diff
@@ -478,3 +480,12 @@ message CompositePriceConfiguration {
   // List of each price source and its corresponding binding
   repeated vega.SpecBindingForCompositePrice data_sources_spec_binding = 8;
 }
+
+// Describes which properties of the data source data is to be
+// used for automated purchase.
+message DataSourceSpecToAutomatedPurchaseBinding {
+  // Name of the property in the source data that should be used to determine the automated purchase schedule.
+  string auction_schedule_property = 1;
+  // Name of the property in the source data that should be used to determine the schedule of .
```
Comment (candida-d) [inline protos/sources/vega/markets.proto:489]:
```suggestion
  // Name of the property in the source data that should be used to determine the schedule of the automated purchase auction.
```

File: protos/sources/vega/events/v1/events.proto [hunk 9]
```diff
@@ -1804,3 +1809,11 @@ message VolumeRebateProgramEnded {
   // Epoch at which the referral program ended.
   uint64 at_epoch = 4;
 }
+
+message AutomatedPurchaseAnnounced {
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1813]: Could you document these fields too?

Comment (ze97286) [reply protos/sources/vega/events/v1/events.proto:1814]: added

File: protos/sources/vega/events/v1/events.proto [hunk 10]
```diff
@@ -1804,3 +1809,17 @@ message VolumeRebateProgramEnded {
   // Epoch at which the referral program ended.
   uint64 at_epoch = 4;
 }
+
+// an event indicating the schedule of an automated purchase auction
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1813]:
```suggestion
// Event indicating the schedule of an automated purchase auction
```

File: protos/sources/vega/events/v1/events.proto [hunk 11]
```diff
@@ -1804,3 +1809,17 @@ message VolumeRebateProgramEnded {
...
+message AutomatedPurchaseAnnounced {
+  // the token being exchanged.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1815]:
```suggestion
  // Token being exchanged.
```

File: protos/sources/vega/events/v1/events.proto [hunk 12]
```diff
@@ -1804,3 +1809,17 @@ message VolumeRebateProgramEnded {
...
+  string from = 1;
+  // the source account type.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1817]:
```suggestion
  // The source account type.
```

File: protos/sources/vega/events/v1/events.proto [hunk 13]
```diff
@@ -1804,3 +1809,17 @@ message VolumeRebateProgramEnded {
...
+  vega.AccountType from_account_type = 2;
+  // the target account type.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1819]:
```suggestion
  // The target account type.
```

File: protos/sources/vega/events/v1/events.proto [hunk 14]
```diff
@@ -1804,3 +1809,17 @@ message VolumeRebateProgramEnded {
...
+  vega.AccountType to_account_type = 3;
+  // the identifier of the spot market where the auction is being scheduled.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1821]:
```suggestion
  // The identifier of the spot market where the auction is being scheduled. 
```

File: protos/sources/vega/events/v1/events.proto [hunk 15]
```diff
@@ -1804,3 +1809,17 @@ message VolumeRebateProgramEnded {
...
+message AutomatedPurchaseAnnounced {
+  // the token being exchanged.
+  string from = 1;
+  // the source account type.
+  vega.AccountType from_account_type = 2;
+  // the target account type.
+  vega.AccountType to_account_type = 3;
+  // the identifier of the spot market where the auction is being scheduled. 
+  string market_id = 4;
+  // the amount being exchanged.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1823]:
```suggestion
  // Amount being exchanged. 
```

File: datanode/gateway/graphql/schema.graphql [hunk 16]
```diff
@@ -7839,6 +7847,35 @@ enum AMMStatusReason {
...
+  priceOracle: DataSourceDefinition!
+  "The final acceptable price is the price oracle value weighted by this factor (e.g. 1.05 to allow for 5% slippage on the purchase)"
+  oracleOffsetFactor: String! 
+  "A time based oracle for when auctions will occur"
+  auctionSchedule: DataSourceDefinition
+  "The duration of the auction"
+  auction_duration: String!
+  "A time based oracle for when an observation will be taken of the balance of the source account. This will emit an event notifying of the balance planned to exchange, along with storing this value. When an auction occurs, the latest reading for this value will be used for the volume to trade, rather than the full balance of the account"
+  auctionVolumeSnapshotSchedule: DataSourceDefinition!
+  "Minimum number of tokens to be sold (specified in asset decimals). If less than this are available in the account at the last snapshot before auction, no auction will occur and the balance will roll over to the next scheduled auction"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:7869]:
```suggestion
  "Minimum number of tokens to be sold (specified in asset decimals). If less than this is available in the account at the last snapshot before auction, no auction will occur and the balance will roll over to the next scheduled auction"
```

File: datanode/gateway/graphql/schema.graphql [hunk 17]
```diff
@@ -7839,6 +7847,35 @@ enum AMMStatusReason {
...
+  minimumAuctionSize: String!
+  "Maximum number of tokens to be sold (specified in asset decimals). If more than this are available in the account at the last snapshot before auction, this maximum value will be used instead, and the remainder will be rolled over to the next scheduled auction"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:7871]:
```suggestion
  "Maximum number of tokens to be sold (specified in asset decimals). If more than this is available in the account at the last snapshot before auction, this maximum value will be used instead, and the remainder will be rolled over to the next scheduled auction"
```

File: datanode/gateway/graphql/schema.graphql [hunk 18]
```diff
@@ -7839,6 +7847,35 @@ enum AMMStatusReason {
...
+  maximumAuctionSize: String!
+  "The expiry timestamp in seconds of the automated purchase. 0 if no expiry configured"
+  expiryTimestamp: Int!
+  "For how long the price from the price oracle is considered usable"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:7875]: What format is this in? Seconds?

File: datanode/gateway/graphql/schema.graphql [hunk 19]
```diff
@@ -7839,6 +7847,35 @@ enum AMMStatusReason {
...
+  "The market which will be used to enact the purchase/sale"
+  marketId: ID
+  "The oracle which will define an approximate acceptable price for the transaction"
+  priceOracle: DataSourceDefinition!
+  "The final acceptable price is the price oracle value weighted by this factor (e.g. 1.05 to allow for 5% slippage on the purchase)"
+  oracleOffsetFactor: String! 
+  "A time based oracle for when auctions will occur"
+  auctionSchedule: DataSourceDefinition
+  "The duration of the auction"
+  auction_duration: String!
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:7866]: Should this be formatted auctionDuration?

File: protos/sources/vega/markets.proto [hunk 20]
```diff
@@ -478,3 +480,12 @@ message CompositePriceConfiguration {
   // List of each price source and its corresponding binding
   repeated vega.SpecBindingForCompositePrice data_sources_spec_binding = 8;
 }
+
+// Describes which properties of the data source data is to be
```
Comment (candida-d) [inline protos/sources/vega/markets.proto:484]:
```suggestion
// Describes which properties of the data source data are to be
```

File: protos/sources/vega/governance.proto [hunk 21]
```diff
@@ -882,3 +888,41 @@ message RecurringTransfer {
...
+  // A time based oracle for when an observation will be taken of the balance of the source account. This will emit an event notifying of the balance planned to exchange, along with storing this value. When an auction occurs, the latest reading for this value will be used for the volume to trade, rather than the full balance of the account.
+  vega.DataSourceDefinition auction_volume_snapshot_schedule = 9;
+  // The spec binding for the auction schedule and volume snapshot schedule
+  vega.DataSourceSpecToAutomatedPurchaseBinding automated_purchase_spec_binding = 10;
+  // The duration of the auction
+  string auction_duration = 11;
+  // Minimum number of tokens to be sold (specified in asset decimals). If less than this are available in the account at the last snapshot before auction, no auction will occur and the balance will roll over to the next scheduled auction.
+  string minimum_auction_size = 12;
+  // Maximum number of tokens to be sold (specified in asset decimals). If more than this are available in the account at the last snapshot before auction, this maximum value will be used instead, and the remainder will be rolled over to the next scheduled auction.
```
Comment (candida-d) [inline protos/sources/vega/governance.proto:922]:
```suggestion
  // Maximum number of tokens to be sold (specified in asset decimals). If more than this is available in the account at the last snapshot before auction, this maximum value will be used instead, and the remainder will be rolled over to the next scheduled auction.
```

File: protos/sources/vega/governance.proto [hunk 22]
```diff
@@ -882,3 +888,41 @@ message RecurringTransfer {
...
+  string oracle_offset_factor = 7;
+  // A time based oracle for when auctions will occur
+  vega.DataSourceDefinition auction_schedule = 8;
+  // A time based oracle for when an observation will be taken of the balance of the source account. This will emit an event notifying of the balance planned to exchange, along with storing this value. When an auction occurs, the latest reading for this value will be used for the volume to trade, rather than the full balance of the account.
+  vega.DataSourceDefinition auction_volume_snapshot_schedule = 9;
+  // The spec binding for the auction schedule and volume snapshot schedule
+  vega.DataSourceSpecToAutomatedPurchaseBinding automated_purchase_spec_binding = 10;
+  // The duration of the auction
+  string auction_duration = 11;
+  // Minimum number of tokens to be sold (specified in asset decimals). If less than this are available in the account at the last snapshot before auction, no auction will occur and the balance will roll over to the next scheduled auction.
```
Comment (candida-d) [inline protos/sources/vega/governance.proto:920]:
```suggestion
  // Minimum number of tokens to be sold (specified in asset decimals). If less than this is available in the account at the last snapshot before auction, no auction will occur and the balance will roll over to the next scheduled auction.
```

File: protos/sources/vega/governance.proto [hunk 23]
```diff
@@ -882,3 +888,41 @@ message RecurringTransfer {
...
+message NewProtocolAutomatedPurchaseChanges {
+  // The source token which will be swapped
+  string from = 1;
+  // The account type for the network from which the tokens will be sourced
+  vega.AccountType from_account_type = 2;
+  // The account type for the network to which the purchased tokens will be sent
+  vega.AccountType to_account_type = 3;
+  // The market which will be used to enact the purchase/sale
+  string market_id = 4;
+  // The oracle which will define an approximate acceptable price for the transaction
```
Comment (candida-d) [inline protos/sources/vega/governance.proto:906]:
```suggestion
  // The oracle that will define an approximate acceptable price for the transaction
```

File: protos/sources/vega/governance.proto [hunk 24]
```diff
@@ -882,3 +888,41 @@ message RecurringTransfer {
...
+}
+
+message NewProtocolAutomatedPurchaseChanges {
+  // The source token which will be swapped
+  string from = 1;
+  // The account type for the network from which the tokens will be sourced
+  vega.AccountType from_account_type = 2;
+  // The account type for the network to which the purchased tokens will be sent
+  vega.AccountType to_account_type = 3;
+  // The market which will be used to enact the purchase/sale
```
Comment (candida-d) [inline protos/sources/vega/governance.proto:904]:
```suggestion
  // The market that will be used to enact the purchase/sale
```

File: protos/sources/vega/governance.proto [hunk 25]
```diff
@@ -882,3 +888,41 @@ message RecurringTransfer {
...
   string factor = 4;
 }
+
+message NewProtocolAutomatedPurchase {
+  // Configuration for a change to create a new automated purchase auction
+  NewProtocolAutomatedPurchaseChanges changes = 1;
+}
+
+message NewProtocolAutomatedPurchaseChanges {
+  // The source token which will be swapped
```
Comment (candida-d) [inline protos/sources/vega/governance.proto:898]:
```suggestion
  // The source token that will be swapped
```

File: datanode/gateway/graphql/schema.graphql [hunk 26]
```diff
@@ -7839,6 +7847,35 @@ enum AMMStatusReason {
   STATUS_REASON_CANNOT_REBASE
 }
 
+type NewProtocolAutomatedPurchase {
+  "The source token which will be swapped"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:7851]:
```suggestion
  "The source token that will be swapped"
```

File: datanode/gateway/graphql/schema.graphql [hunk 27]
```diff
@@ -7839,6 +7847,35 @@ enum AMMStatusReason {
...
+  from: ID!
+  "The account type for the network from which the tokens will be sourced"
+  fromAccountType: AccountType!
+  "The account type for the network to which the purchased tokens will be sent"
+  toAccountType: AccountType!
+  "The market which will be used to enact the purchase/sale"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:7857]:
```suggestion
  "The market that will be used to enact the purchase/sale"
```

File: datanode/gateway/graphql/schema.graphql [hunk 28]
```diff
@@ -7839,6 +7847,35 @@ enum AMMStatusReason {
...
+  marketId: ID
+  "The oracle which will define an approximate acceptable price for the transaction"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:7859]:
```suggestion
  "The oracle that will define an approximate acceptable price for the transaction"
```

File: datanode/gateway/graphql/schema.graphql (hunk 19 above)
Comment (ze97286) [reply datanode/gateway/graphql/schema.graphql:7866]: yes, I'll fix

File: datanode/gateway/graphql/schema.graphql (hunk 18 above)
Comment (ze97286) [reply datanode/gateway/graphql/schema.graphql:7875]: in seconds yes

File: datanode/gateway/graphql/schema.graphql [hunk 29]
```diff
@@ -7846,6 +7854,35 @@ enum AMMStatusReason {
...
+  auctionDuration: String!
+  "A time based oracle for when an observation will be taken of the balance of the source account. This will emit an event notifying of the balance planned to exchange, along with storing this value. When an auction occurs, the latest reading for this value will be used for the volume to trade, rather than the full balance of the account"
+  auctionVolumeSnapshotSchedule: DataSourceDefinition!
+  "Minimum number of tokens to be sold (specified in asset decimals). If less than this is available in the account at the last snapshot before auction, no auction will occur and the balance will roll over to the next scheduled auction"
+  minimumAuctionSize: String!
+  "Maximum number of tokens to be sold (specified in asset decimals). If more than this is available in the account at the last snapshot before auction, this maximum value will be used instead, and the remainder will be rolled over to the next scheduled auction"
+  maximumAuctionSize: String!
+  "The expiry timestamp in seconds of the automated purchase. 0 if no expiry configured"
+  expiryTimestamp: Int!
+  "For how long the price from the price oracle is considered usable"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:7882]:
```suggestion
  "For how long the price from the price oracle is considered usable, in seconds"
```

**Review by candida-d — DISMISSED** [review]

**Review by candida-d — DISMISSED** [review]

File: core/execution/spot/market.go [hunk 30]
```diff
@@ -3387,3 +3487,60 @@ func (m *Market) CheckOrderSubmissionForSpam(orderSubmission *types.OrderSubmiss
...
+		os := &types.OrderSubmission{
+			MarketID:    m.mkt.ID,
+			Price:       orderPrice.Clone(),
+			Size:        orderSize,
+			Side:        orderSide,
+			TimeInForce: types.OrderTimeInForceGFA,
+			Type:        types.OrderTypeLimit,
+			Reference:   reference,
+		}
+		conf, err := m.SubmitOrder(ctx, os, types.NetworkParty, crypto.RandomHash())
```
Comment (wwestgarth) [inline core/execution/spot/market.go:3532]: Is `crypto.RandomHash()` here not going to cause snapshot issues, or events being sent to the data-node being different?

Comment (ze97286) [reply core/execution/spot/market.go:3532]: why snapshot issues - the order id is persistent
why being sent to different data nodes?

Comment (wwestgarth) [reply core/execution/spot/market.go:3532]: `crypto.RandomHash()` will end up being used to generate the order ID no? Which will be different across different nodes. It'll get saved into a snapshot if one happens during the auction period, and also any event the data-node sends out for that order will have the random ID in it.

Comment (ze97286) [reply core/execution/spot/market.go:3532]: I forgot the random hash doesn't take have a seed, will fix it.

File: core/events/bus.go [hunk 31]
```diff
@@ -383,6 +385,7 @@ var (
 		VolumeRebateProgramEndedEvent:            eventspb.BusEventType_BUS_EVENT_TYPE_VOLUME_REBATE_PROGRAM_ENDED,
 		VolumeRebateProgramUpdatedEvent:          eventspb.BusEventType_BUS_EVENT_TYPE_VOLUME_REBATE_PROGRAM_UPDATED,
 		VolumeRebateStatsUpdatedEvent:            eventspb.BusEventType_BUS_EVENT_TYPE_VOLUME_REBATE_STATS_UPDATED,
+		AutomatedPurchaseAnnouncedEvent:          eventspb.BusEventType_BUS_EVENT_TYPE_AUTOMATED_PURCHASE_ANNOUNCED,
```
Comment (wwestgarth) [inline core/events/bus.go:388]: I think you also need to add this to `datanode/broker/convert.go`

**Review by wwestgarth — DISMISSED** [review]

**Review by wwestgarth — APPROVED** [review]

**Review by candida-d — APPROVED** [review]

**Review by cdummett — APPROVED** [review]

**Pull request merged**